source/_posts/analyzing-javascript-with-babel-traverse.md

```diff
@@ -0,0 +1,182 @@
+- get a list of the styles within that stylesheet
+- go through and find all references to that stylesheet, e.g. `myStyleSheet.someStyle`
+- from that, determine which styles are never referenced and can be safely deleted
+
+#### Parse the JavaScript file
+
+We'll be using the libraries [babylon]() and [babel-traverse](). The [plugins section](https://github.com/thejameskyle/babel-handbook/blob/master/translations/en/plugin-handbook.md) of [@thejameskyle](https://twitter.com/thejameskyle)'s [babel-handbook](https://github.com/thejameskyle/babel-handbook) is also an excellect reference.
+
+```js
+const babylon = require('babylon')
+const text = fs.readFileSync(file, 'utf8')
+const ast = babylon.parse(text, {
+  // this means that `import` and `export` are allowed
+  sourceType: 'module',
+  // we want to allow all the fancy syntax
+  plugins: ['jsx', 'flow', 'objectRestSpread', 'classProperties'],
+})
+```
+
+#### Find the StyleSheet declarations
+of the form `const myStyleSheet = StyleSheet.create({ .... })`
+
+We're going to use [`babel-traverse`](https://github.com/babel/babel/tree/master/packages/babel-traverse), which makes walking through the tree super easy.
+
+It gives us a function, `traverse`, that will walk through the whole AST, and call the visitors we specify corresponding to the type of a given node. To figure out what the types are (and what the shape of the AST is like), I lean on [astexplorer.net](https://astexplorer.net/) and [the Readme](https://github.com/babel/babel/blob/master/packages/babel-types/README.md) of [babel-types](https://github.com/babel/babel/tree/master/packages/babel-types).
+
+```js
+const styleSheets = []
+traverse(ast, {
+  CallExpression(path) {
+    if (isStyleSheetCreate(path)) {
+      const members = path.node.arguments[0].properties.filter(
+        property => property.type === 'ObjectProperty' // Not gonna try to figure out spreads
+      )
+      const styleNames = members.map(member => member.key.name)
+      styleSheets.push({id: path.parent.id, styleNames})
+    }
+  },
+})
+```
+
+So I go through every [`CallExpression`](https://github.com/babel/babel/blob/master/packages/babel-types/README.md#callexpression), which looks like `something(arg1, arg2, ...)`, check if it looks like `StyleSheet.create`, and then process it. The function `isStyleSheetCreate` (that I'll describe in a second) determines whether the current node looks like `var myStyleSheet = StyleSheet.create({y: ... })`. I then grab all the members of the object that's being passed to `StyleSheet.create`, discarding any that happen to be [`ObjectMethod`](https://github.com/babel/babel/blob/master/packages/babel-types/README.md#objectmethod)s or [`computed`](https://github.com/babel/babel/blob/master/packages/babel-types/README.md#objectproperty) properties, and get the names that they're being identified by. In our example file, `styleNames` would end up being `["header", "awesome"]`.
```

Comment:
Maybe: "which has the form `something(arg1, arg2, ...)`, check if it looks like..." or something similar, to avoid back-to-back "looks like".